File: java/src/io/agora/media/DynamicKey6.java
package io.agora.media;

import org.apache.commons.codec.binary.Base64;
import org.apache.commons.codec.binary.Hex;

import java.util.Date;
import java.util.Random;
import java.util.TreeMap;

/**
 * Created by liwei on 7/9/17.
 */
public class DynamicKey6 {
    public final static String version = "006";
    public final static String noUpload = "0";
    public final static String audioVideoUpload = "3";

    // ServiceType
    public final static short MEDIA_CHANNEL_SERVICE = 1;
    public final static short RECORDING_SERVICE = 2;
    public final static short PUBLIC_SHARING_SERVICE = 3;
    public final static short IN_CHANNEL_PERMISSION = 4;

    // InChannelPermissionKey
    public final static short ALLOW_UPLOAD_IN_CHANNEL = 1;

    private static final int VERSION_LENGTH = 3;
    private static final int APP_ID_LENGTH = 32;

    static String generateSignature(String appCertificate, short service, String appID, int unixTs, int salt, String channelName, long uid, int expiredTs, TreeMap<Short, String> extra) throws Exception {
        // decode hex to avoid case problem
        Hex hex = new Hex();
        byte[] rawAppID = hex.decode(appID.getBytes());
        byte[] rawAppCertificate = hex.decode(appCertificate.getBytes());

        Message m = new Message(service, rawAppID, unixTs, salt, channelName, (int)(uid & 0xFFFFFFFFL), expiredTs, extra);
        byte[] toSign = pack(m);
        return new String(Hex.encodeHex(DynamicKeyUtil.encodeHMAC(rawAppCertificate, toSign), false));
    }

    public static String generateDynamicKey(String appID, String appCertificate, String channel, int ts, int salt, long uid, int expiredTs, TreeMap<Short, String> extra, short service) throws Exception {
        String signature = generateSignature(appCertificate, service, appID, ts, salt, channel, uid, expiredTs, extra);
        DynamicKey6Content content = new DynamicKey6Content(service, signature, (int)uid, ts, salt, expiredTs, extra);
        byte[] bytes = pack(content);
        byte[] encoded = new Base64().encode(bytes);
        String base64 = new String(encoded);
        return version + appID + base64;
    }

    private static byte[] pack(Packable content) {
        ByteBuf buffer = new ByteBuf();
        content.marshall(buffer);
        return buffer.asBytes();
    }

    public static String generatePublicSharingKey(String appID, String appCertificate, String channel, int ts, int salt, long uid, int expiredTs) throws Exception {
        return generateDynamicKey(appID, appCertificate, channel, ts, salt, uid, expiredTs, new TreeMap<Short, String>(), PUBLIC_SHARING_SERVICE);
    }

    public static String generateRecordingKey(String appID, String appCertificate, String channel, int ts, int salt, long uid, int expiredTs) throws Exception {
        return generateDynamicKey(appID, appCertificate, channel, ts, salt, uid, expiredTs, new TreeMap<Short, String>(), RECORDING_SERVICE);
    }

    public static String generateMediaChannelKey(String appID, String appCertificate, String channel) throws Exception {
        return generateMediaChannelKey(appID, appCertificate, channel, 0);
    }

    public static String generateMediaChannelKey(String appID, String appCertificate, String channel, long uid) throws Exception {
        return generateMediaChannelKey(appID, appCertificate, channel, uid, 0);
    }

    public static String generateMediaChannelKey(String appID, String appCertificate, String channel, long uid, int expiredTs) throws Exception {
        return generateMediaChannelKey(appID, appCertificate, channel, getTimestamp(), getSalt(), uid, expiredTs);
    }

    public static String generateMediaChannelKey(String appID, String appCertificate, String channel, int ts, int salt, long uid, int expiredTs) throws Exception {
        return generateDynamicKey(appID, appCertificate, channel, ts, salt, uid, expiredTs, new TreeMap<Short, String>(), MEDIA_CHANNEL_SERVICE);
    }

    public static String generateInChannelPermissionKey(String appID, String appCertificate, String channel, long uid, String permission) throws Exception {
        return generateInChannelPermissionKey(appID, appCertificate, channel, getTimestamp(), getSalt(), uid, 0, permission);
    }

    public static String generateInChannelPermissionKey(String appID, String appCertificate, String channel, int ts, int salt, long uid, int expiredTs, String permission) throws Exception {
        TreeMap<Short, String> extra = new TreeMap<Short, String>();
        extra.put(ALLOW_UPLOAD_IN_CHANNEL, permission);
        return generateDynamicKey(appID, appCertificate, channel, ts, salt, uid, expiredTs, extra, IN_CHANNEL_PERMISSION);
    }

    String signature = "";
    String appID = "";
    long uid;
    int unixTs ;
    int salt;
    int expiredTs;
    TreeMap<Short, String> extra = new TreeMap<Short, String>();

    static int getTimestamp() {
        return (int)(new Date().getTime()/1000);
    }

    private static int getSalt() {
        return new Random(new Date().getTime()).nextInt();
    }

    public boolean fromString(String channelKeyString) {
        if (!version.equals(channelKeyString.substring(0, VERSION_LENGTH))) {
            return false;
        }

        appID = channelKeyString.substring(VERSION_LENGTH, VERSION_LENGTH + APP_ID_LENGTH);
        if (appID.length() != APP_ID_LENGTH) {
            return false;
        }

        byte[] rawContent = new Base64().decode(channelKeyString.substring(VERSION_LENGTH + APP_ID_LENGTH));
        if (rawContent.length == 0) {
            return false;
        }

        DynamicKey6Content content = new DynamicKey6Content(rawContent);
        signature = content.signature;
        uid = content.uid & 0xFFFFFFFFL;
        unixTs = content.unixTs;
        salt = content.salt;
        expiredTs = content.expiredTs;
        extra = content.extra;
        return true;
    }

    static class Message implements Packable {
        public short serviceType;
        public byte[] appID;
        public int unixTs;
        public int salt;
        public String channelName;
        public int uid;
        public int expiredTs;
        public TreeMap<Short, String> extra;

        public Message(short serviceType, byte[] appID, int unixTs, int salt, String channelName, int uid, int expiredTs, TreeMap<Short, String> extra) {
            this.serviceType = serviceType;
            this.appID = appID;
            this.unixTs = unixTs;
            this.salt = salt;
            this.channelName = channelName;
            this.uid = uid;
            this.expiredTs = expiredTs;
            this.extra = extra;
        }

        public ByteBuf marshall(ByteBuf out) {
            return out.put(serviceType).put(appID).put(unixTs).put(salt).put(channelName).put(uid).put(expiredTs).put(extra);
        }
    }

    static class DynamicKey6Content implements Packable {
        public short serviceType;
        public String signature = "";
        public int uid;
        public int unixTs;
        public int salt;
        public int expiredTs;
        public TreeMap<Short, String> extra = new TreeMap<Short, String>();

        public DynamicKey6Content(short serviceType, String signature, int uid, int unixTs, int salt, int expiredTs, TreeMap<Short, String> extra) {
            this.serviceType = serviceType;
            this.signature = signature;
            this.uid = uid;
            this.unixTs = unixTs;
            this.salt = salt;
            this.expiredTs = expiredTs;
            this.extra = extra;
        }

        public DynamicKey6Content(byte[] bytes) {
            unmarshall(new ByteBuf(bytes));
        }

        public ByteBuf marshall(ByteBuf out) {
            return out.put(serviceType).put(signature).put(uid).put(unixTs).put(salt).put(expiredTs).put(extra);
        }

        public void unmarshall(ByteBuf in) {
            this.serviceType = in.readShort();
            this.signature = in.readString();
            this.uid = in.readInt();
            this.unixTs = in.readInt();
            this.salt = in.readInt();
            this.expiredTs = in.readInt();
            in.readMap(this.extra);
        }
    }}

File: nodejs/src/DynamicKey6.js
var crypto = require('crypto');

var version = "006";
var noUpload = "0";
var audioVideoUpload = "3";

var generatePublicSharingKeyFull = function (appID, appCertificate, channelName, unixTs, randomInt, uid, expiredTs) {
    channelName=channelName.toString();
    return generateDynamicKey(appID, appCertificate, channelName, unixTs, randomInt, uid, expiredTs, null, PUBLIC_SHARING_SERVICE);
};

var generateRecordingKeyFull = function (appID, appCertificate, channelName, unixTs, randomInt, uid, expiredTs) {
    channelName=channelName.toString();
    return generateDynamicKey(appID, appCertificate, channelName, unixTs, randomInt, uid, expiredTs, null, RECORDING_SERVICE);
};

var generateMediaChannelKeyFull = function (appID, appCertificate, channelName, unixTs, randomInt, uid, expiredTs) {
    channelName=channelName.toString();
    return generateDynamicKey(appID, appCertificate, channelName, unixTs, randomInt, uid, expiredTs, null, MEDIA_CHANNEL_SERVICE);
};

var generateInChannelPermissionKeyFull = function (appID, appCertificate, channelName, unixTs, randomInt, uid, expiredTs, permission) {
    var extra = {};
    extra[ALLOW_UPLOAD_IN_CHANNEL] = permission;
    return generateDynamicKey(appID, appCertificate, channelName, unixTs, randomInt, uid, expiredTs, extra, IN_CHANNEL_PERMISSION);
};

var generateDynamicKey = function (appID, appCertificate, channelName, unixTs, randomInt, uid, expiredTs, extra, serviceType) {
    var signature = generateSignature6(appCertificate, serviceType, appID, unixTs, randomInt, channelName, uid, expiredTs, extra);
    var content = DynamicKey6Content({
        serviceType: serviceType
        , signature: signature
        , uid: uid
        , unixTs: unixTs
        , salt: randomInt
        , expiredTs: expiredTs
        , extra: extra}).pack();
    return version + appID + content.toString('base64');
};

var fromString = function(keyString) {
    var that = {
        parsed: false
        , signature: ""
        , appID: ""
        , uid: 0
        , unixTs: 0
        , salt: 0
        , expiredTs: 0
        , extra: {}
    };

    if (keyString.substr(0, VERSION_LENGTH) != version) {
        return that;
    }

    var appID = keyString.substr(VERSION_LENGTH, APP_ID_LENGTH);
    if (appID.length != APP_ID_LENGTH) {
        return that;
    }
    that.appID = appID;

    var rawContent = Buffer.from(keyString.substr(VERSION_LENGTH + APP_ID_LENGTH), 'base64');
    if (rawContent.length == 0) {
        return that;
    }

    var content = DynamicKey6Content().unpack(rawContent);

    that.signature = content.signature;
    that.uid = content.uid;
    that.unixTs = content.unixTs;
    that.salt = content.salt;
    that.expiredTs = content.expiredTs;
    that.extra = content.extra;

    that.parsed = true;

    return that;
};

var generateSignature6 = function(appCertificate, serviceType, appID, unixTs, randomInt, channelName, uid, expiredTs, extra) {
    // decode hex to avoid case problem
    var rawAppID = hexDecode(appID);
    var rawAppCertificate = hexDecode(appCertificate);

    var m = Message({
        serviceType: serviceType
        , appID: rawAppID
        , unixTs: unixTs
        , salt: randomInt
        , channelName: channelName
        , uid: uid
        , expiredTs: expiredTs
        , extra: extra
    });

    var toSign = m.pack();
    return encodeHMac(rawAppCertificate, toSign);
};

module.exports.version = version;
module.exports.noUpload = noUpload;
module.exports.audioVideoUpload = audioVideoUpload;
module.exports.generatePublicSharingKeyFull = generatePublicSharingKeyFull;
module.exports.generateRecordingKeyFull = generateRecordingKeyFull;
module.exports.generateMediaChannelKeyFull = generateMediaChannelKeyFull;
module.exports.generateInChannelPermissionKeyFull = generateInChannelPermissionKeyFull;
module.exports.generateDynamicKey = generateDynamicKey;
module.exports.generateSignature = generateSignature6;
module.exports.fromString = fromString;

var encodeHMac = function(key, message) {
    return crypto.createHmac('sha1', key).update(message).digest('hex').toUpperCase();
};

var hexDecode = function(str) {
    return new Buffer(str, 'hex');
};

var ByteBuf = function() {
    var that = {
        buffer: new Buffer(1024)
        , position: 0
    };

    that.buffer.fill(0);

    that.pack = function() {
        var out = new Buffer(that.position);
        that.buffer.copy(out, 0, 0, out.length);
        return out;
    };

    that.putUint16 = function(v) {
        that.buffer.writeUInt16LE(v, that.position);
        that.position += 2;
        return that;
    };

    that.putUint32 = function(v) {
        that.buffer.writeUInt32LE(v, that.position);
        that.position += 4;
        return that;
    };

    that.putBytes = function(bytes) {
        that.putUint16(bytes.length);
        bytes.copy(that.buffer, that.position);
        that.position += bytes.length;
        return that;
    };

    that.putString = function(str) {
        return that.putBytes(new Buffer(str));
    };

    that.putTreeMap = function(map) {
        if (!map) {
            that.putUint16(0);
            return that;
        }

        that.putUint16(Object.keys(map).length);
        for (var key in map) {
            that.putUint16(key);
            that.putString(map[key]);
        }

        return that;
    };

    that.getUint16 = function() {
        var v = that.buffer.readUInt16LE(that.position);
        that.position += 2;
        return v;
    };

    that.getUint32 = function() {
        var v = that.buffer.readUInt32LE(that.position);
        that.position += 4;
        return v;
    };

    that.getString = function() {
        var length = that.getUint16();
        var str = that.buffer.toString('ascii', that.position, length + that.position);
        that.position += length;
        return str;
    };

    that.getTreeMap = function() {
        var m = {};

        var count = that.getUint16();
        for (var i = 0; i < count; ++i) {
            var k = that.getUint16();
            var v = that.getString();
            m[k] = v;
        }

        return m;
    };

    return that;
}

var DynamicKey6Content = function(options) {
    var that = {};
    that.pack = function() {
        var out = ByteBuf();
        return out.putUint16(options.serviceType)
            .putString(options.signature)
            .putUint32(options.uid)
            .putUint32(options.unixTs)
            .putUint32(options.salt)
            .putUint32(options.expiredTs)
            .putTreeMap(options.extra)
            .pack();
    };

    that.unpack = function(buffer) {
        var buf = ByteBuf();
        buf.buffer = buffer;

        var content = {};
        content.serviceType = buf.getUint16();
        content.signature = buf.getString();
        content.uid = buf.getUint32();
        content.unixTs = buf.getUint32();
        content.salt = buf.getUint32();
        content.expiredTs = buf.getUint32();
        content.extra = buf.getTreeMap();
        return content;
    };

    return that;
}

var Message = function(options) {
    options.pack = function() {
        var out = ByteBuf();
        return out.putUint16(options.serviceType)
            .putBytes(options.appID)
            .putUint32(options.unixTs)
            .putUint32(options.salt)
            .putString(options.channelName)
            .putUint32(options.uid)
            .putUint32(options.expiredTs)
            .putTreeMap(options.extra)
            .pack();
    }

    return options;
}

// InChannelPermissionKey
var ALLOW_UPLOAD_IN_CHANNEL = 1;

// Service Type
var MEDIA_CHANNEL_SERVICE = 1;
var RECORDING_SERVICE = 2;
var PUBLIC_SHARING_SERVICE = 3;
var IN_CHANNEL_PERMISSION = 4;

// length
var VERSION_LENGTH = 3;
var APP_ID_LENGTH = 32;
